### PR TITLE
Add a runtime parameter to PODVector and DynamicBitset - whether to pin memory

### PR DESCRIPTION
**Pull request by srogatch**
https://github.com/KatanaGraph/katana-enterprise/pull/1551 depends on this PR.
JIRA issue: https://katanagraph.atlassian.net/browse/KAT-908

### CODE REVIEW

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: I don't like `Usual`. What about `NotPinned`?

Also let's not limit the size unless we actually expect it to matter. This is a premature optimization that can result in ABI compatibility issues.

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: Why not `MemoryPinType`?

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: Please handle errors in both `mlock` and `munlock`, throughout this code. It might be reasonable to just print a warning (to alert people of the performance issue), but the error shouldn't just be ignored. We may also need to prevent `munlock` from being called if `mlock` fails.

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: This is always used.

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: I think these should be `KATANA_LOG_ASSERT` the reason is that we are already allocating which is slow and catching allocation failures quickly in some form in release builds is important. Since it can happen if the user doesn't have enough RAM in their system.

File: libsupport/include/katana/CpuGpuSwitch.h
Comment (arthurp) [inline libsupport/include/katana/CpuGpuSwitch.h]: This isn't used yet. I would rather add this file in a later PR that actually uses it.

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h]: This duplicates the move constructor (and the default constructor if that exists). Maybe factor it out into a private `Clear` method.

File: libgalois/include/katana/PODVector.h
Comment (arthurp) [inline libgalois/include/katana/PODVector.h:74]: You are removing the default constructor. I think the compiler might be generating one, but it might be better to be explicit. I think (though I'm not sure) that an explicit constructor which can be called with zero-arguments via a default parameter is *different* from a default constructor (not marked explicit and with zero-arguments natively).

File: libgalois/include/katana/DynamicBitset.h
Comment (arthurp) [inline libgalois/include/katana/DynamicBitset.h]: Same default constructor concern.

**Review by arthurp — CHANGES_REQUESTED** [review]
Good over all. I'd like to clean some things up though.

**Review by amberhassaan — CHANGES_REQUESTED** [review]
Hi @srogatch : 
I think we can solve this problem more cleanly with "allocators".  Just like `std::vector<T, alloctor>`, we should allow `PODVector` and `DynmaicBitSet` to take `PinnedMemoryAllocator` (and other allocators, like GPUallocator if it makes sense). This will remove need for `if(pinned_)...` type code and contain the complexity. 
The approach below is not scalable complexity wise and we should be investing into better memory management abstractions.

File: libgalois/include/katana/PODVector.h
Comment (amberhassaan) [reply libgalois/include/katana/PODVector.h]: We can enhance it further and include GPU memory in MemoryTypes and I agree with Arthur that `Usual` should be `NonPinned`

**Comment by srogatch** [comment]
> I think we can solve this problem more cleanly with "allocators". Just like `std::vector<T, alloctor>`, we should allow `PODVector` and `DynmaicBitSet` to take `PinnedMemoryAllocator` (and other allocators, like GPUallocator if it makes sense). This will remove need for `if(pinned_)...` type code and contain the complexity.
> The approach below is not scalable complexity wise and we should be investing into better memory management abstractions.

@amberhassaan , introducing an allocator is really a lot of work and changes to the code. `DynamicBitset` is not a template and it has functions with external linkage. The latter is solvable by providing externally-linked specializations, however, the former requires changes all over the code adding something like `<>` if we provide a default template parameter for the allocator. Furthermore, I've already tried to make the `pinned` flag a template parameter to `DynamicBitset` and `PODVector` and got stuck. We would need to make a lot of other functions templates, many of them are externally linked, in order to accept both `DynamicBitset<true>` and `DynamicBitset<false>` (or the equivalents with allocators) so to be able to copy data between them. The same for usages of `PODVector`: we want to enable copying between pinned and non-pinned `PODVector`s and we want functions that work with both. Finally, we want to switch between pinned and non-pinned vectors/bitsets at runtime, while a template parameter would force us to make this decision at compile-time. Perhaps I cannot foresee every problem with introducing the allocator, but it should be already clear that this change is very intrusive and requires many changes to the external code.

**Comment by amberhassaan** [comment]
> > I think we can solve this problem more cleanly with "allocators". Just like `std::vector<T, alloctor>`, we should allow `PODVector` and `DynmaicBitSet` to take `PinnedMemoryAllocator` (and other allocators, like GPUallocator if it makes sense). This will remove need for `if(pinned_)...` type code and contain the complexity.
> > The approach below is not scalable complexity wise and we should be investing into better memory management abstractions.
> 
> @amberhassaan , introducing an allocator is really a lot of work and changes to the code. `DynamicBitset` is not a template and it has functions with external linkage. The latter is solvable by providing externally-linked specializations, however, the former requires changes all over the code adding something like `<>` if we provide a default template parameter for the allocator. Furthermore, I've already tried to make the `pinned` flag a template parameter to `DynamicBitset` and `PODVector` and got stuck. We would need to make a lot of other functions templates, many of them are externally linked, in order to accept both `DynamicBitset<true>` and `DynamicBitset<false>` (or the equivalents with allocators) so to be able to copy data between them. The same for usages of `PODVector`: we want to enable copying between pinned and non-pinned `PODVector`s and we want functions that work with both. Finally, we want to switch between pinned and non-pinned vectors/bitsets at runtime, while a template parameter would force us to make this decision at compile-time. Perhaps I cannot foresee every problem with introducing the allocator, but it should be already clear that this change is very intrusive and requires many changes to the external code.

Hi @srogatch , there's 1) misunderstanding 2) strong disagreement (on my side). 
Miunderstanding: I'm not asking for a `bool isPinned` as a template flag to DynamicBitSet or PODVector. Rather, I'm asking for a separation of concerns. Allocators deal with how to allocated memory. Pinned or not-pinned, CPU or GPU. PODVector or DynamicBItset (or any data structures) shouldn't have complex logic and flags of this sort. Please take a look at STL containers, all of which allow passing a custom allocator object. 
AFA the change due to adding an allocator as template parameter goes, we can limit that with typedefs (or type aliases), e.g. 
```
template <typename T, typename AllocatorType>
class PODVectorImplementation {
...
};

template <typename T>
using PODVector = PODVectorImplementation<T, SomeDefaultAllocator>;

template <typename T>
using PODVectorPinned = PODVectorImplementation<T, PinnedAllocator>;
```
And, you can add copy constructors between PODVector and PODVector Pinned. STL containers like `std::vector` already do this.

**Comment by srogatch** [comment]
@amberhassaan , there is no misunderstanding on my side: I understood about the allocator, but it's a huge amount of (re)work. However, it seems you missed my point and I am not sure what to do except to ask you to re-read my comment. Whether it's allocator or a bool in a template parameter is irrelevant. The problem is that there is a huge amount of conversion in the code between pinned and non-pinned `PODVector` and `DynamicBitset`, and if they are templated, the conversion would be problematic. The details are in my previous comment. I strongly disagree with the allocator approach, and I can only suggest to create a JIRA issue so that you can look yourself into the opportunity of changing to the allocator. I tried adding a template parameter to `PODVector` and `DynamicBitset`, so I know what I am talking about. The amount of (re)work is huge.

**Comment by srogatch** [comment]
Created https://katanagraph.atlassian.net/browse/KAT-949 .

File: libgalois/include/katana/PODVector.h
Comment (srogatch) [reply libgalois/include/katana/PODVector.h]: How about `Swappable`? I try to avoid negations in the naming and the code, just think of `if(x != MemoryPinType::NotPinned)`. Pinned memory cannot go to a swap file, in contrast to swappable memory.

File: libsupport/include/katana/CpuGpuSwitch.h
Comment (srogatch) [reply libsupport/include/katana/CpuGpuSwitch.h]: This is currently used in https://github.com/KatanaGraph/katana-enterprise/pull/1551 . I had to place it in `katana` rather than `katana-enterprise` repo because someone may use the constants for `DynamicBitset` and `PODVector` instances in `katana` repo too.

File: libgalois/include/katana/PODVector.h
Comment (srogatch) [reply libgalois/include/katana/PODVector.h:231]: I've tested and such constructors seem to behave exactly the same. Without the no-argument constructor, the code compiles well, including the places that use `PODVector` with default construction (e.g. inside an `std::vector`). However, if I add an explicit no-argument constructor, I get compilation errors: `error: call of overloaded ‘PODVector()’ is ambiguous`. So there should be no problem here with the code as it is.

File: libgalois/include/katana/PODVector.h
Comment (srogatch) [reply libgalois/include/katana/PODVector.h:74]: I've tested and such constructors seem to behave exactly the same. Without the no-argument constructor, the code compiles well, including the places that use `PODVector` with default construction (e.g. inside an `std::vector`). However, if I add an explicit no-argument constructor, I get compilation errors: `error: call of overloaded ‘PODVector()’ is ambiguous`. So there should be no problem here with the code as it is.

File: libgalois/include/katana/DynamicBitset.h
Comment (srogatch) [reply libgalois/include/katana/DynamicBitset.h]: Should be ok.

**Comment by amberhassaan** [comment]
> @amberhassaan , there is no misunderstanding on my side: I understood about the allocator, but it's a huge amount of (re)work. However, it seems you missed my point and I am not sure what to do except to ask you to re-read my comment. Whether it's allocator or a bool in a template parameter is irrelevant. The problem is that there is a huge amount of conversion in the code between pinned and non-pinned `PODVector` and `DynamicBitset`, and if they are templated, the conversion would be problematic. The details are in my previous comment. I strongly disagree with the allocator approach, and I can only suggest to create a JIRA issue so that you can look yourself into the opportunity of changing to the allocator. I tried adding a template parameter to `PODVector` and `DynamicBitset`, so I know what I am talking about. The amount of (re)work is huge.

HI @srogatch , if you think adding a template parameter for allocator induces a lot of work, we can consider an alternative, where we define a hierarchy of allocators with virtual functions (Pinned and NonPinned allocators being derived classes). Typically, with these containers (like PODVector), allocations are not frequent. 

My main point is that we need better memory management abstractions than having flags and calling variants of malloc based on `if conditions`. Going the allocator route (or a similar abstraction), we can make it more sophisticated going forward. See https://github.com/LLNL/CHAI and https://github.com/LLNL/Umpire

**Comment by srogatch** [comment]
> HI @srogatch , if you think adding a template parameter for allocator induces a lot of work, we can consider an alternative, where we define a hierarchy of allocators with virtual functions (Pinned and NonPinned allocators being derived classes). Typically, with these containers (like PODVector), allocations are not frequent.
> 
> My main point is that we need better memory management abstractions than having flags and calling variants of malloc based on `if conditions`. Going the allocator route (or a similar abstraction), we can make it more sophisticated going forward. See https://github.com/LLNL/CHAI and https://github.com/LLNL/Umpire

Hi @amberhassaan ! The polymorphism-based allocators may take a reasonable amount of effort, but I'm not sure. An investigation is needed. I would still like to keep it out of the scope of this PR and https://github.com/KatanaGraph/katana-enterprise/pull/1551 , because it's not GPU/CUDA-related and I just added pinning/unpinning of memory to the places which already use `calloc()`/`malloc()`/`realloc()`/`free()` or `new`/`delete`. I also fixed obvious memory leaks following the rule to improve along the way. But RAII and Allocators would take more work time than pinning/unpinning itself. And no customer pays for code quality: they pay for correctness, features, and performance. That is why we should keep the quality improvement efforts adequate to the number of changes we make for features, performance, and bug-fixing. I'm quite new to the project, so I could be most affected by code quality, but I wouldn't say that the code quality is so bad that we have to spend 3x time on technical debt compared to feature/performance/bug-fixing.

**Comment by amberhassaan** [comment]
@srogatch : We shouldn't be cutting corners on code quality because it always comes back to bite you. The cost and effort we spend on it pays for it self through long term maintainability and robustness (avoid and catching bugs early). Bad code, weak abstractions affect every developer, so it's men * hours.

**Review by amberhassaan — APPROVED** [review]

**Comment by amberhassaan** [comment]
I'm OK with adding TODOs along the lines of needing allocators.

**Comment by srogatch** [comment]
Thanks, @amberhassaan . I've just added TODOs, also to https://github.com/KatanaGraph/katana-enterprise/pull/1551 . The RAII task seems to be the next on my list: https://katanagraph.atlassian.net/browse/KAT-952 .

**Comment by arthurp** [comment]
> I wouldn't say that the code quality is so bad that we have to spend 3x time on technical debt compared to feature/performance/bug-fixing.

The CUDA code in this code base really is that bad. I agree that such improvements should be a separate PR. But I do not agree that we should limit code quality improvements so we can focus on features/performance/whatever. These code quality issues will bite us. And it may not take long as @amberhassaan said. Please try to find some time to do these refactors. They are important.

**Comment by arthurp** [comment]
As a side note, I would MUCH prefer a dynamic polymorphic allocator to a template parameter. In our use cases allocation is rare and fairly slow, so the cost of a virtual function call will be low. And using a dynamic allocator parameter will make integrating with Python and potentially even the allocators of other python libraries much easier.

File: libsupport/include/katana/CpuGpuSwitch.h
Comment (arthurp) [inline libsupport/include/katana/CpuGpuSwitch.h]: Responded at: https://github.com/KatanaGraph/katana/pull/379#discussion_r673352128

In the end, I'll leave it to you. But if there is no current need for this in open, I strongly recommend not including it.

File: libsupport/include/katana/CpuGpuSwitch.h
Comment (arthurp) [reply libsupport/include/katana/CpuGpuSwitch.h]: I don't really understand this argument. These values only have meaning for GPU enabled builds, and they should really be dynamic parameters or allocators anyway not build parameters, so hopefully we will be removing them in a refactor soon. So I see no reason to allow people to use them in more places that we will have to fix later.

**Review by arthurp — APPROVED** [review]
Seems OK. Please address my last comment before you merge. And we'll talk about refactors soon.

**Comment by srogatch** [comment]
> Seems OK. Please address my last comment before you merge. And we'll talk about refactors soon.

Thanks. I've started work on the polymorphic pinned allocator: https://github.com/KatanaGraph/katana/pull/383 , https://github.com/KatanaGraph/katana-enterprise/pull/1562 .

I'll now switch to this PR to finalize.

**Pull request merged**
